### PR TITLE
Native transactions implementation

### PR DESCRIPTION
**Pull request by vkagamlyk**
Current state of PR:

- [x] CRUD with transactions
- [x] Unit tests for `TinkerTransactionGraph`
- [x] Gherkin test setup
- [x] `TinkerHelper` refactoring
- [x] Handle for indices
- [x] Test fixes
- [x] Performance test results
- [x] Documentation

### CODE REVIEW

File: gremlin-server/src/test/scripts/generate-all.groovy [hunk 1]
```diff
@@ -36,7 +36,7 @@ globals << [hook : [
     // things now, that test config would have been mixed in with release artifacts and there would have been ugly
     // exclusions to make packaging work properly.
     allowSetOfIdManager = { graph, idManagerFieldName, idManager ->
-        java.lang.reflect.Field idManagerField = graph.class.getDeclaredField(idManagerFieldName)
+        java.lang.reflect.Field idManagerField = graph.class.getSuperclass().getDeclaredField(idManagerFieldName)
```
Comment (Cole-Greer) [inline gremlin-server/src/test/scripts/generate-all.groovy:39]: I'm not very familiar with this script, is `graph` in this context guaranteed to be an instance of either TinkerGraph or TinkerTransactionGraph? Is it possible that this may be used by other providers and inserting `getSuperclass()` will break their testing?

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/algorithm/generator/AbstractGeneratorTest.java [hunk 2]
```diff
@@ -49,6 +49,7 @@ protected boolean same(final Graph g1, final Graph g2) {
                         final Iterator<Vertex> innerItty = g2.vertices();
                         final Vertex v = IteratorUtils.filter(innerItty, vx -> vx.value("oid").equals(p.getValue0())).next();
                         CloseableIterator.closeIterator(innerItty);
+                        final boolean b = sameInVertices(v, p.getValue1()) && sameOutVertices(v, p.getValue2());
```
Comment (Cole-Greer) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/algorithm/generator/AbstractGeneratorTest.java:52]: Is this variable ever used?

Comment (vkagamlyk) [reply gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/algorithm/generator/AbstractGeneratorTest.java:52]: it was debug code, removed

File: gremlin-server/src/test/scripts/generate-all.groovy (hunk 1 above)
Comment (vkagamlyk) [reply gremlin-server/src/test/scripts/generate-all.groovy:39]: yes, other providers should use own scripts

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/FeatureSupportTest.java [hunk 3]
```diff
@@ -121,7 +121,8 @@ public void shouldSupportComputerIfAGraphCanCompute() throws Exception {
          */
         @Test
         @FeatureRequirement(featureClass = GraphFeatures.class, feature = FEATURE_TRANSACTIONS, supported = false)
-        public void shouldSupportTransactionsIfAGraphConstructsATx() throws Exception {
+        @FeatureRequirement(featureClass = GraphFeatures.class, feature = FEATURE_THREADED_TRANSACTIONS, supported = false)
+        public void shouldSupportTransactionsIfAGraphConstructsATx() {
             try {
                 graph.tx();
                 fail(String.format(INVALID_FEATURE_SPECIFICATION, GraphFeatures.class.getSimpleName(), FEATURE_TRANSACTIONS));
```
Comment (Cole-Greer) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/FeatureSupportTest.java:128]: Should `FEATURE_TRANSACTIONS` be updated here to `FEATURE_THREADED_TRANSACTIONS`?

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerVertex.java [hunk 4]
```diff
@@ -27,14 +27,7 @@
...
 
-import java.util.ArrayList;
-import java.util.Collections;
-import java.util.HashMap;
-import java.util.Iterator;
-import java.util.List;
-import java.util.Map;
-import java.util.Optional;
-import java.util.Set;
+import java.util.*;
```
Comment (Cole-Greer) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerVertex.java:30]: Prefer not to use wildcard imports

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerVertex.java [hunk 5]
```diff
@@ -43,17 +36,40 @@
 public final class TinkerVertex extends TinkerElement implements Vertex {
 
     protected Map<String, List<VertexProperty>> properties;
-    protected Map<String, Set<Edge>> outEdges;
-    protected Map<String, Set<Edge>> inEdges;
-    private final TinkerGraph graph;
+    protected Map<String, Set<Object>> outEdges;
```
Comment (Cole-Greer) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerVertex.java:39]: Do the edge maps need to be relaxed to Object types here?

Comment (vkagamlyk) [reply tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerVertex.java:39]: yes, because map stores edge id

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/FeatureSupportTest.java (hunk 3 above)
Comment (vkagamlyk) [reply gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/FeatureSupportTest.java:128]: no, `FEATURE_THREADED_TRANSACTIONS` means transaction can be executed across multiple threads

**Review by xiazcy — COMMENTED** [review]
Just some documentation nits

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/io/GraphMigrator.java [hunk 6]
```diff
@@ -37,7 +37,8 @@
  */
 public final class GraphMigrator {
 
-    private GraphMigrator() {}
+    private GraphMigrator() {
```
Comment (spmallette) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/io/GraphMigrator.java:40]: nit: revert spacing changes?

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/CollectionUtil.java [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+ */
+
+package org.apache.tinkerpop.gremlin.util;
+
+import java.util.ArrayList;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+public final class CollectionUtil {
```
Comment (spmallette) [inline gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/CollectionUtil.java:27]: there is a `org.apache.tinkerpop.gremlin.util.tools` with a `CollectionFactory`. should we move the functions there to this class and remove `CollectionFactory`? i think this is the better named class. may as well get rid of "tools" package and just move `MultiMap` up a level to this package. please `final` variables in this class and add unit tests for `clone()` - any reason the method uses a capital "C" and not lower case?

File: docs/src/upgrade/release-3.7.x.asciidoc [hunk 8]
```diff
@@ -60,6 +60,54 @@ gremlin> g.union(V().has('name','vadas'),
 
 See: link:https://issues.apache.org/jira/browse/TINKERPOP-2873[TINKERPOP-2873]
 
+==== Native transaction support
```
Comment (spmallette) [inline docs/src/upgrade/release-3.7.x.asciidoc:63]: A few things regarding the upgrade docs:

1. I think you should title this as "TinkerGraph Transactions" because we didn't really implemented "native transactions" in TinkerPop to apply to any graph. They are transactions specific to TinkerGraph itself as a reference implementation.
2. I'm not sure what's written here is telling the right story for the Upgrade Documentation. It's sort of focused on Gremlin Server with examples that really don't have much to do with the specific change that you implemented. I think you want to explain (a) that TinkerGraph has transaction functionality now, (b) summarize its capability briefly, (c) include Console session output for how to create a transactional TinkerGraph, add a vertex, commit/rollback (d) explain when best to use it.

File: gremlin-server/src/test/scripts/test-server-start.groovy [hunk 9]
```diff
@@ -60,7 +60,10 @@ settings.graphs.modern = platformAgnosticGremlinServerDir + "/src/test/scripts/t
 settings.graphs.crew = platformAgnosticGremlinServerDir + "/src/test/scripts/tinkergraph-empty.properties"
 settings.graphs.grateful = platformAgnosticGremlinServerDir + "/src/test/scripts/tinkergraph-empty.properties"
 settings.graphs.sink = platformAgnosticGremlinServerDir + "/src/test/scripts/tinkergraph-empty.properties"
-if (testTransactions) settings.graphs.tx = platformAgnosticGremlinServerDir + "/src/test/scripts/neo4j-empty.properties"
+if (testTransactions) {
+    settings.graphs.tx = platformAgnosticGremlinServerDir + "/src/test/scripts/neo4j-empty.properties"
```
Comment (spmallette) [inline gremlin-server/src/test/scripts/test-server-start.groovy:64]: is there any point to configuring neo4j for testing now that we have TinkerGraph? can't we just remove the configuration? iirc we only have neo4j here to validate that remote transactions work on the server and with drivers and we can now do that with TinkerGraph. we really don't need to test that neo4j itself works in this context.

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/AbstractGraphProvider.java [hunk 10]
```diff
@@ -151,5 +151,7 @@ protected void readIntoGraph(final Graph graph, final String path) throws IOExce
         final String dataFile = TestHelper.generateTempFileFromResource(graph.getClass(),
                 GryoResourceAccess.class, path.substring(path.lastIndexOf(File.separator) + 1), "", false).getAbsolutePath();
         graph.traversal().io(dataFile).read().iterate();
+        if (graph.features().graph().supportsTransactions())
```
Comment (spmallette) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/AbstractGraphProvider.java:154]: i'm pretty sure we left this code out purposefully and meant for the graph to handle commit after read on its own. unless there's a good reason to change this, i'd say it would be better to move this code to TinkerGraph's test implementation and just add to the javadoc to note that implementers should handle commit on their own.

File: gremlin-util/src/test/java/org/apache/tinkerpop/gremlin/util/ser/GraphSONMessageSerializerV3Test.java [hunk 11]
```diff
@@ -288,9 +288,9 @@ public void shouldSerializeToTreeJson() throws Exception {
         Iterator<Vertex> vertexKeys = firstTree.keySet().iterator();
 
         Map expectedValues = new HashMap<Integer, String>() {{
-            put(3, "vadas");
-            put(5, "lop");
-            put(7, "josh");
+            put(2, "vadas");
```
Comment (spmallette) [inline gremlin-util/src/test/java/org/apache/tinkerpop/gremlin/util/ser/GraphSONMessageSerializerV3Test.java:291]: what made all ids change (i see you had to do this in other tests as well)?

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/AbstractTinkerGraph.java [hunk 12]
```diff
@@ -0,0 +1,508 @@
...
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.stream.Stream;
+
+public abstract class AbstractTinkerGraph implements Graph {
```
Comment (spmallette) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/AbstractTinkerGraph.java:49]: please add some javadoc for the class (though it might be nice to add some more throughout). also, missing some `final` designations.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/AbstractTinkerIndex.java [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+
+package org.apache.tinkerpop.gremlin.tinkergraph.structure;
+
+import org.apache.tinkerpop.gremlin.structure.Element;
+
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+public abstract class AbstractTinkerIndex<T extends Element> {
```
Comment (spmallette) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/AbstractTinkerIndex.java:28]: javadoc please...

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerElementContainer.java [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.tinkerpop.gremlin.tinkergraph.structure;
+
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.locks.ReentrantLock;
+
+final class TinkerElementContainer<T extends TinkerElement> {
```
Comment (spmallette) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerElementContainer.java:24]: won't call out any more javadoc items, but i think it's worth a review of new classes/new transaction methods in this PR to document them up.

File: gremlin-server/src/test/scripts/test-server-start.groovy (hunk 9 above)
Comment (mikepersonick) [reply gremlin-server/src/test/scripts/test-server-start.groovy:64]: Yes, part of this work should be pulling out the Neo4j integration

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/algorithm/generator/AbstractGeneratorTest.java [hunk 15]
```diff
@@ -58,11 +58,23 @@ protected boolean same(final Graph g1, final Graph g2) {
 
     private boolean sameInVertices(final Vertex v, final List<Vertex> inVertices) {
         return inVertices.stream()
-                .allMatch(inVertex -> IteratorUtils.filter(v.vertices(Direction.IN), hv -> hv.value("oid").equals(inVertex.value("oid"))).hasNext());
+                .allMatch(inVertex ->
+                {
+                    final Iterator<Vertex> itty = v.vertices(Direction.IN);
+                    final boolean result = IteratorUtils.filter(itty, hv -> hv.value("oid").equals(inVertex.value("oid"))).hasNext();
+                    CloseableIterator.closeIterator(itty);
```
Comment (mikepersonick) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/algorithm/generator/AbstractGeneratorTest.java:65]: Can we do this in a try/finally everywhere so we don't leak resources in exception cases?

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/algorithm/generator/AbstractGeneratorTest.java [hunk 16]
```diff
@@ -58,11 +58,23 @@ protected boolean same(final Graph g1, final Graph g2) {
...
     }
 
     private boolean sameOutVertices(final Vertex v, final List<Vertex> outVertices) {
         return outVertices.stream()
-                .allMatch(outVertex -> IteratorUtils.filter(v.vertices(Direction.OUT), hv -> hv.value("oid").equals(outVertex.value("oid"))).hasNext());
+                .allMatch(outVertex ->
+                {
+                    final Iterator<Vertex> itty = v.vertices(Direction.OUT);
+                    final boolean result = IteratorUtils.filter(itty, hv -> hv.value("oid").equals(outVertex.value("oid"))).hasNext();
+                    CloseableIterator.closeIterator(itty);
```
Comment (mikepersonick) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/algorithm/generator/AbstractGeneratorTest.java:76]: same

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/StructureStandardSuite.java [hunk 17]
```diff
@@ -105,6 +105,7 @@ public class StructureStandardSuite extends AbstractGremlinSuite {
             SerializationTest.class,
             StarGraphTest.class,
             TransactionTest.class,
+            TransactionTestV2.class,
```
[the remark this answers is not in the document]
Comment (mikepersonick) [reply gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/StructureStandardSuite.java:108]: If you needed additional test coverage just add it to TransactionTest. "V2" is not a good name for a test.

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/TransactionTestV2.java [hunk 18]
```diff
@@ -0,0 +1,671 @@
...
+import static org.apache.tinkerpop.gremlin.structure.Graph.Features.EdgeFeatures.FEATURE_ADD_EDGES;
+import static org.apache.tinkerpop.gremlin.structure.Graph.Features.EdgeFeatures.FEATURE_REMOVE_EDGES;
+import static org.apache.tinkerpop.gremlin.structure.Graph.Features.ElementFeatures.FEATURE_ADD_PROPERTY;
+import static org.apache.tinkerpop.gremlin.structure.Graph.Features.GraphFeatures.FEATURE_TRANSACTIONS;
+import static org.apache.tinkerpop.gremlin.structure.Graph.Features.VertexFeatures.*;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class TransactionTestV2 extends AbstractGremlinTest {
```
Comment (mikepersonick) [inline gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/TransactionTestV2.java:36]: Either move these back into TransactionTest (if they are general tests that apply to any provider) or move them into the Tinkergraph module (if they are specific tests for your implementation). Either way, we also need a TinkerGraphTransactionStressTest that stress tests your implementation under higher concurrency. We didn't have stress tests for Neo4j and other provider transactions because we assumed that the provider already had those natively and we could trust that their transactions implementation worked - we can no longer assume that. Stress test meaning many threads trying to do many thing concurrently and making sure that the outcome is what we'd expect. That should probably be in the Tinkergraph module though.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerHelper.java [hunk 19]
```diff
@@ -53,176 +49,102 @@ public final class TinkerHelper {
...
-    }
-
-    public static void removeIndex(final TinkerEdge edge, final String key, final Object value) {
-        final TinkerGraph graph = (TinkerGraph) edge.graph();
-        if (graph.edgeIndex != null)
-            graph.edgeIndex.remove(key, value, edge);
-    }
-
+    // todo: move to TinkerVertex?
     public static Iterator<TinkerEdge> getEdges(final TinkerVertex vertex, final Direction direction, final String... edgeLabels) {
```
Comment (mikepersonick) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerHelper.java:74]: Why were these changes necessary?

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransaction.java [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+            txChangedVertices.get().add((TinkerElementContainer<TinkerVertex>) container);
+        } else {
+            if (null == txChangedEdges.get())
+                txChangedEdges.set(new HashSet<>());
+            txChangedEdges.get().add((TinkerElementContainer<TinkerEdge>) container);
+        }
+    }
+
+    @Override
+    protected void doCommit() throws TransactionException {
```
Comment (mikepersonick) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransaction.java:94]: I think we need a narrative here (Javadoc) explaining the thought process and approach. It's a bit tough to decipher just looking at the code and it will be hard for others to work on this code later.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransaction.java [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+
+            changedVertices.forEach(v -> v.releaseLock());
+            changedEdges.forEach(e -> e.releaseLock());
+
+            txNumber.set(NOT_STARTED);
+        }
+    }
+
+    @Override
+    protected void doRollback() throws TransactionException {
```
Comment (mikepersonick) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransaction.java:152]: same as above

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerElementContainer.java (hunk 14 above)
Comment (mikepersonick) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerElementContainer.java:24]: Same as the Transaction class, I think we need a narrative here (Javadoc) explaining the thought process and approach. It's a bit tough to decipher just looking at the code and it will be hard for others to work on this code later. markDeleted -> touch -> setDraft -> touch ... some explanation would go a long way.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraph.java [hunk 22]
```diff
@@ -0,0 +1,499 @@
...
+ *
+ * @author Valentyn Kahamlyk
+ */
+@Graph.OptIn(Graph.OptIn.SUITE_STRUCTURE_STANDARD)
+@Graph.OptIn(Graph.OptIn.SUITE_STRUCTURE_INTEGRATE)
+@Graph.OptIn(Graph.OptIn.SUITE_PROCESS_STANDARD)
+@Graph.OptIn(Graph.OptIn.SUITE_PROCESS_COMPUTER)
+@Graph.OptIn(Graph.OptIn.SUITE_PROCESS_LIMITED_STANDARD)
+@Graph.OptIn(Graph.OptIn.SUITE_PROCESS_LIMITED_COMPUTER)
+public final class TinkerTransactionGraph extends AbstractTinkerGraph {
```
Comment (mikepersonick) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraph.java:59]: It would be great if we could get rid of AbstractTinkerGraph. Just make TinkerGraph the base class and TinkerTransactionGraph extend it. Not sure if it's possible.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionalIndex.java [hunk 23]
```diff
@@ -0,0 +1,239 @@
...
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * @author Marko A. Rodriguez (http://markorodriguez.com)
```
Comment (mikepersonick) [inline tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionalIndex.java:33]: ?

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerVertex.java (hunk 5 above)
Comment (mikepersonick) [reply tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerVertex.java:47]: Can we document this in the code? I assume this is related to that test case change? Why was it necessary?

File: tinkergraph-gremlin/src/test/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraphTest.java [hunk 24]
```diff
@@ -0,0 +1,722 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.assertNull;
+import static org.junit.Assert.fail;
+
+public class TinkerTransactionGraphTest {
```
Comment (mikepersonick) [inline tinkergraph-gremlin/src/test/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraphTest.java:38]: stress tests

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/CollectionUtil.java (hunk 7 above)
Comment (vkagamlyk) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/util/CollectionUtil.java:32]: consolidated in https://github.com/apache/tinkerpop/pull/2087/commits/9b6991dd123267ae5ffa536cb608e81d4c999aad

File: gremlin-util/src/test/java/org/apache/tinkerpop/gremlin/util/ser/GraphSONMessageSerializerV3Test.java (hunk 11 above)
Comment (vkagamlyk) [reply gremlin-util/src/test/java/org/apache/tinkerpop/gremlin/util/ser/GraphSONMessageSerializerV3Test.java:291]: the main reason is the change in the processing order of the vertices. Because of this, the vertex properties were created in a different order and received different IDs. Property IDs are not defined anywhere so I left it as is

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/util/Attachable.java [hunk 25]
```diff
@@ -197,8 +197,10 @@ public static Optional<Edge> getEdge(final Attachable<Edge> attachableEdge, fina
             final Iterator<Edge> edgeIterator = hostVertex.edges(Direction.OUT, attachableEdge.get().label());
             while (edgeIterator.hasNext()) {
                 final Edge edge = edgeIterator.next();
-                if (ElementHelper.areEqual(edge, baseEdge))
+                if (ElementHelper.areEqual(edge, baseEdge)) {
+                    CloseableIterator.closeIterator(edgeIterator);
```
[the remark this answers is not in the document]
Comment (vkagamlyk) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/util/Attachable.java:201]: Yes, this is a problem, but I think it should be solved also in 3.5/3.6.
I added [ticket](https://issues.apache.org/jira/browse/TINKERPOP-2970) to track this.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerFactory.java [hunk 26]
```diff
@@ -54,10 +51,21 @@ public static TinkerGraph createClassic() {
         return g;
     }
 
+    /**
+     * Create the "classic" graph with transaction support.
+     * @see #createClassic()
+     */
+    public static TinkerTransactionGraph createTxClassic() {
```
[the remark this answers is not in the document]
Comment (kenhuuu) [reply tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerFactory.java:58]: It was done this way so that users could easily get either a transaction version or regular version of TinkerGraph without having to supply their own graph instance. This helps to keep it easy to use. Also, since there is currently only two implementations, I felt it was OK to just add Tx versions rather than completely making separate Factory classes for each of them. Though, I would agree that it makes sense to change how this works in the future if an implementation was added or removed.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraph.java (hunk 22 above)
Comment (vkagamlyk) [reply tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraph.java:59]: I think technically it's possible, but
1. Internal implementation related to elements storage is quite different, so overrided code will be a mess. 
2. I think at some point difference between `TinkerGraph` and `TinkerTransactionGraph` can become bigger, for example for performance is better to use different structures for Vertex and Edge, not same like now.

File: tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerHelper.java (hunk 19 above)
Comment (vkagamlyk) [reply tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerHelper.java:74]: I changed the way how vertices contains linked edges. The main reason is the ability to set a transaction boundary.

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/StructureStandardSuite.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (vkagamlyk) [reply gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/structure/StructureStandardSuite.java]: I renamed it and added explanation

File: gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/AbstractGraphProvider.java (hunk 10 above)
Comment (vkagamlyk) [reply gremlin-test/src/main/java/org/apache/tinkerpop/gremlin/AbstractGraphProvider.java]: addressed in https://github.com/apache/tinkerpop/pull/2087/commits/595d8273753c1d4bf3ed072cf9afd5215dcc0420

File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/io/GraphMigrator.java (hunk 6 above)
Comment (vkagamlyk) [reply gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/structure/io/GraphMigrator.java]: done

File: tinkergraph-gremlin/src/test/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraphTest.java (hunk 24 above)
Comment (kenhuuu) [reply tinkergraph-gremlin/src/test/java/org/apache/tinkerpop/gremlin/tinkergraph/structure/TinkerTransactionGraphTest.java:38]: Performance and stress tests that were run against TinkerTransactionGraph can be found [here](https://github.com/Bit-Quill/tinkerpop/blob/ken/txntester/tinkergraph-gremlin/src/main/java/org/apache/tinkerpop/gremlin/tinkergraph/Tester.java). In particular, the functional testing portion of that Tester was used to discover multithreaded issues caused by running many transactions on the same Element.

**Comment by vkagamlyk** [comment]
VOTE +1

**Comment by Cole-Greer** [comment]
LGTM, thanks Valentyn it will be great to have this added support in TinkerGraph. VOTE +!

**Comment by xiazcy** [comment]
LGTM, VOTE +1. Thanks Valentyn!
